Break up dependency management into individual chapters

bmuschko opened the pull request:
### Context

Promotes better readability and is now available in the left navigation bar. See https://github.com/gradle/dotorg-docs/issues/133 for more information.

Additional comments:

- I renamed the "Dependency Management Basics" to "Dependency Management for Java Projects". IMO that best resembles the current content. I put that under "Managing Dependencies > Tutorials" for now. There're still some pieces in there that are not covered otherwise so we'll have to transfer them in some form before we can delete it. Alternatively, I could envision this page to stage as a more trimmed downed version. Other languages, domains (e.g. native, Android) could live there side-by-side. TBD!
- In the single page documentation I creates a new subsection for "Dependency Management". As far as I could see the infrastructure doesn't allow subsections yet.
- I see the chapters "Dependency types" and "Repository types" as reference material. That's why it lives under "Managing Dependencies > Glossary" right now. Happy to discuss other options.

General infrastructure issues I encountered:

- Asciidoc files do not get processed properly if they are put in a subdirectory of `src/docs/userguide`.
- The DocBook XSL processing seems to strip out certain portions of a file name e.g. the `dependencyManagement` prefix. Maybe it tries to make them as short as possible? It does keep the prefix though if there's a potential conflict e.g. `introduction.html` vs. `dependency_management_introduction.html`.

eriwen commented on subprojects/docs/src/docs/userguide/dependencyManagementWorkingWithDependencies.adoc:
```diff
@@ -0,0 +1,87 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+[[working_with_dependencies]]
```
:x: Note that the filename `dependencyManagementWorkingWithDependencies.adoc` needs to match the generated URL (currently `working_with_dependencies.html`) otherwise this will break the "Edit this page" widget.

bmuschko replied:
done

eriwen commented on subprojects/docs/src/docs/userguide/dependencyManagementTroubleshootingDependencyResolution.adoc:
```diff
@@ -0,0 +1,97 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+[[troubleshooting_dependency_resolution]]
```
:x: Note that the filename needs to match the generated URL otherwise this will break the "Edit this page" widget.

bmuschko replied:
done

eriwen commented on subprojects/docs/src/docs/userguide/dependencyManagementTroubleshootingDependencyResolution.adoc:
```diff
@@ -0,0 +1,97 @@
+// limitations under the License.
+
+[[troubleshooting_dependency_resolution]]
+== Troubleshooting dependency resolution
```
Please use Title Case for page titles only.

bmuschko replied:
done

eriwen commented on subprojects/docs/src/docs/userguide/dependencyManagementWorkingWithDependencies.adoc:
```diff
@@ -0,0 +1,87 @@
+// limitations under the License.
+
+[[working_with_dependencies]]
+== Working with dependencies
```
Please use Title Case for page titles

bmuschko replied:
done

eriwen commented on subprojects/docs/src/docs/userguide/dependencyManagementWorkingWithDependencies.adoc:
```diff
@@ -0,0 +1,87 @@
+[[working_with_dependencies]]
+== Working with dependencies
+
+For the examples below we have the following dependencies setup:
```
Where does "Working with Dependencies" fit into your [IA plan](https://docs.google.com/document/d/1op4_f_TC3ESJO7xvtvu0mrLpucTMJD8uVINqZMHVfdI/edit#heading=h.qbe1l85zx8r7)?

bmuschko replied:
That's the chapter I mentioned that we need to rewrite completely. I'd probably put it under "Inspecting Dependencies" depending on the content. At first I was tempted to just delete it but there're aspects that people want to know about. I will create a new issue for it.

eriwen replied:
Okay. This definitely needs some love.

bmuschko replied:
The content we have here right now needs to be completely scrapped from my perspective.

eriwen commented on subprojects/docs/src/docs/userguide/dependencyManagementTroubleshootingDependencyResolution.adoc:
```diff
@@ -0,0 +1,97 @@
+[[troubleshooting_dependency_resolution]]
+== Troubleshooting dependency resolution
+
+Managing large dependency graphs can be challenging. This section describes techniques for troubleshooting issues you might encounter in your project.
```
Is it in the short term plan to update this content?

bmuschko replied:
Depends on how much time we'll want to spend on the overall dependency management chapter. I think rewriting the introduction is much more important. We have an issue for this portion on the board. From my perspective we need to touch all chapters really before calling it done.

eriwen commented on subprojects/docs/src/docs/userguide/dependencyManagementManagingTransitiveDependencies.adoc:
```diff
@@ -0,0 +1,129 @@
+// limitations under the License.
+
+[[managing_transitive_dependencies]]
+== Managing transitive dependencies
```
Same comments about filename vs URL and title case as elsewhere

bmuschko replied:
done

eriwen commented on subprojects/docs/src/docs/userguide/dependencyManagementIntroduction.adoc:
```diff
@@ -0,0 +1,117 @@
+// limitations under the License.
+
+[[dependency_management_introduction]]
+== Introduction
```
"Introduction to Dependency Management" would be a clearer, more searchable title I reckon.

eriwen commented on subprojects/docs/src/docs/userguide/dependencyManagementInspectingDependencies.adoc:
```diff
@@ -0,0 +1,85 @@
+// limitations under the License.
+
+[[inspecting_dependencies]]
+== Inspecting dependencies
```
Same comments about file name and title case

bmuschko replied:
done

eriwen commented on subprojects/docs/src/docs/userguide/dependencyManagementDeclaringRepositories.adoc:
```diff
@@ -0,0 +1,75 @@
+// limitations under the License.
+
+[[declaring_repositories]]
+== Declaring repositories
```
Same comment as elsewhere about title case and filename

bmuschko replied:
done

eriwen commented on subprojects/docs/src/docs/userguide/dependencyManagementDeclaringDependencies.adoc:
```diff
@@ -0,0 +1,195 @@
+// limitations under the License.
+
+[[declaring_dependencies]]
+== Declaring dependencies
```
Same comment about filename and title case

bmuschko replied:
done

eriwen commented on subprojects/docs/src/docs/userguide/dependencyManagementCustomizingDependencyResolutionBehavior.adoc:
```diff
@@ -0,0 +1,308 @@
+// limitations under the License.
+
+[[customizing_dependency_resolution_behavior]]
+== Customizing the dependency resolution behavior
```
Same comment about filename and title case

bmuschko replied:
done

eriwen commented on subprojects/docs/src/docs/stylesheets/userGuideHtml.xsl:
```diff
@@ -171,10 +171,25 @@
+                                            <li><a href="/userguide/artifact_dependencies_tutorial.html">Dependency Management for Java Projects</a></li>
+                                        </ul>
+                                    </li>
+                                    <li><a href="/userguide/dependency_management_introduction.html">Introduction</a></li>
```
Consider "Introduction to Dependency Management" perhaps.

bmuschko replied:
done.

eriwen commented on subprojects/docs/src/docs/stylesheets/userGuideHtml.xsl:
```diff
@@ -171,10 +171,25 @@
+                                    <li><a href="/userguide/declaring_repositories.html">Declaring Repositories</a></li>
+                                    <li><a href="/userguide/inspecting_dependencies.html">Inspecting Dependencies</a></li>
+                                    <li><a href="/userguide/managing_transitive_dependencies.html">Managing Transitive Dependencies</a></li>
+                                    <li><a href="/userguide/working_with_dependencies.html">Working with dependencies</a></li>
```
Use title case in the navigation

bmuschko replied:
done

eriwen commented on subprojects/docs/src/docs/stylesheets/userGuideHtml.xsl:
```diff
@@ -171,10 +171,25 @@
-                                    <li><a href="/userguide/repository_types.html">Repository Types</a></li></ul></li>
+                                    <li><a class="nav-dropdown" data-toggle="collapse" href="#dependency-management-tutorials" aria-expanded="false" aria-controls="dependency-management-tutorials">Tutorials</a>
+                                        <ul id="dependency-management-tutorials">
+                                            <li><a href="/userguide/artifact_dependencies_tutorial.html">Dependency Management for Java Projects</a></li>
```
This might belong better under "JVM Projects" bit of navigation. WDYT?

bmuschko replied:
Yes and no. At the moment the page is a mixture between "explaining what dependency management is" and giving some Java-based examples. I think it could fit underneath either section. How about we rewrite the "Introduction to Dependency Management" first and then we what we can remove from the page? But I don't feel too strongly about where it should live right now. At least under "JVM Projects" we already have a Tutorials section.

eriwen commented on subprojects/docs/src/docs/stylesheets/userGuideHtml.xsl:
```diff
@@ -171,10 +171,25 @@
+                                    <li><a href="/userguide/troubleshooting_dependency_resolution.html">Troubleshooting Dependency Resolution</a></li>
+                                    <li><a class="nav-dropdown" data-toggle="collapse" href="#dependency-management-glossary" aria-expanded="false" aria-controls="dependency-management-glossary">Glossary</a>
+                                        <ul id="dependency-management-glossary">
+                                            <li><a href="/userguide/dependency_types.html">Dependency Types</a></li>
```
One really glaring omission in this whole thing is "Configurations". I would consider a single "Concepts" page instead of splitting this into multiple pages. This could be embedded in the introduction.

bmuschko replied:
We actually already have coverage on "configurations". It's just not labeled that way. https://docs.gradle.org/4.5-rc-1/userguide/dependency_management.html#sub:scope_of_dependency_configurations

In a way a "configuration" is a concept but only really makes sense in the context of declaring a dependency. I guess we have the following options:

1. Leave the content where it is with no change. The introduction will explain what a "configuration" is and link to the existing section.
2. Move the content into a completely new page similar to "Dependency types" and "Repository types".
2. Leave the content where it is and change the title to something with "configuration".

eriwen commented on subprojects/docs/src/docs/userguide/dependencyManagement.adoc:
```diff
@@ -1,1006 +0,0 @@
-// See the License for the specific language governing permissions and
-// limitations under the License.
-
-[[dependency_management]]
```
There are surely a lot of links to dependency-management.html. We must decide where we should redirect this URL to if we're not going to keep *some* page at that location.

bmuschko replied:
I guess redirecting to the "Introduction to Dependency Management" would be good enough. Which links are you referring to specifically? You mean in external blog post, guides etc?

bmuschko commented on subprojects/docs/src/docs/userguide/introductionDependencyManagement.adoc:
```diff
@@ -0,0 +1,117 @@
+[[introduction_dependency_management]]
+== Introduction to Dependency Management
+
+Dependency management is a critical feature of every build, and Gradle has placed an emphasis on offering first-class dependency management that is both easy to understand and compatible with a wide variety of approaches. If you are familiar with the approach used by either Maven or Ivy you will be delighted to learn that Gradle is fully compatible with both approaches in addition to being flexib ...
```
The introduction as it stands right now is existing content (it was just moved in this change). I am currently in the process to rewrite it. I am not going to change anything here right now.

eriwen replied:
Got it

bmuschko commented on subprojects/docs/src/docs/userguide/customizingDependencyResolutionBehavior.adoc:
```diff
@@ -0,0 +1,308 @@
+// limitations under the License.
+
+[[customizing_dependency_resolution_behavior]]
+== Customizing the Dependency Resolution Behavior
```
done

bmuschko commented:
> Right, that's what I was thinking. Part of my suggestion is because the new title now is "Dependency Management for Java Projects"

done

> Option 3 might be good. What we want is a section that users will easily find to answer the searches: "What is a configuration?" and "How and why do I manage configurations?"

I changed the title and added additional titles for subsections underneath.

bmuschko commented on subprojects/docs/src/docs/userguide/introductionDependencyManagement.adoc:
```diff
@@ -0,0 +1,117 @@
+
+** When the dependency is declared by a static version and a module descriptor file is found in a repository, there is no need to continue searching later repositories and the remainder of the process is short-circuited.
+
+* All of the artifacts for the module are then requested from the _same repository_ that was chosen in the process above.
```
Seems like it only applies to plugin resolution and their transitive dependencies. I am not sure. I will ask the dependency management team to review the upcoming "introduction" PR.

> Gradle Plugin resolution changes
Plugin resolution now takes all plugin repositories into account and can resolve transitive plugin dependencies across them. Previous versions of Gradle required that all transitive dependencies of a given plugin were present in the same repository as the plugin.

bmuschko commented on subprojects/docs/src/docs/userguide/introductionDependencyManagement.adoc:
```diff
@@ -0,0 +1,117 @@
+
+* Given a required dependency, Gradle first attempts to resolve the _module_ for that dependency. Each repository is inspected in order, searching first for a _module descriptor_ file (POM or Ivy file) that indicates the presence of that module. If no module descriptor is found, Gradle will search for the presence of the primary _module artifact_ file indicating that the module exists in the repo ...
+
+** If the dependency is declared as a dynamic version (like `1.+`), Gradle will resolve this to the newest available static version (like `1.2`) in the repository. For Maven repositories, this is done using the `maven-metadata.xml` file, while for Ivy repositories this is done by directory listing.
```
I believe we mean "highest" here.

bmuschko commented on subprojects/docs/src/docs/userguide/introductionDependencyManagement.adoc:
```diff
@@ -0,0 +1,117 @@
+
+The Gradle dependency cache consists of 2 key types of storage:
+
+* A file-based store of downloaded artifacts, including binaries like jars as well as raw downloaded meta-data like POM files and Ivy files. The storage path for a downloaded artifact includes the SHA1 checksum, meaning that 2 artifacts with the same name but different content can easily be cached.
```
Will do.

bmuschko commented on subprojects/docs/src/docs/userguide/introductionDependencyManagement.adoc:
```diff
@@ -0,0 +1,117 @@
+[[sub:cache_metadata]]
+==== Separate metadata cache
+
+Gradle keeps a record of various aspects of dependency resolution in binary format in the metadata cache. The information stored in the metadata cache includes:
```
The exact location (directory name) of the metadata cache can change by version. That's why I wouldn't want to point to it directly.

bmuschko commented:
@eriwen Your comments on the "introduction" docs have been addressed here: https://github.com/gradle/gradle/pull/4060.
